### PR TITLE
(proof-new) Proof step buffer utilities

### PR DESCRIPTION
**Pull request by ajreynol**
This class is used for delaying proof node creation until it is necessary.

### CODE REVIEW

File: src/expr/proof_step_buffer.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  ProofStep(PfRule r,
+            const std::vector<Node>& children,
+            const std::vector<Node>& args);
+  /** The proof rule */
+  PfRule d_rule;
+  /** The proof children */
+  std::vector<Node> d_children;
+  /** The proof arguments */
+  std::vector<Node> d_args;
+};
```
Comment (HanielB) [inline src/expr/proof_step_buffer.h:42]: Why isn't the conclusion of the step part of the proof step? In the usage below it seems everything would just be simpler if this was the case.

Comment (ajreynol) [reply src/expr/proof_step_buffer.h:42]: I'd like to separate premises from conclusion, a proof step is everything that is necessary for deriving a conclusion.

Comment (ajreynol) [reply src/expr/proof_step_buffer.h:42]: In particular, one reason for this separation: I am often mapping Node (conclusions) to their corresponding ProofStep.  Having conclusion within the ProofStep in this case is redundant.

Comment (HanielB) [reply src/expr/proof_step_buffer.h:42]: OK, that sounds good. Could you just add this motivation for the doc of the step? Might be good to have this motivation said explicitly.

**Pull request merged**